### PR TITLE
Js transport bc

### LINKED ISSUE #414: combined BC for TransportDG
- [x] Implement total_flux and diffusive_flux BC in TransporDG
see #367 
- [x] Allow possibly different BC types in Heat and Concentration model
- [x] Improve description of BC types in the input, it already support formulas in form (($latex formula$)), 
check e.g. html result with 'make html-doc'

### PR DESCRIPTION
**Pull request by jstebel**
Proposal of transport and heat boundary conditions. If OK then some tests need to be added.

Resolves #414.

### CODE REVIEW

File: src/transport/heat_model.cc [hunk 1]
```diff
@@ -42,22 +42,79 @@ using namespace std;
 using namespace Input::Type;
 
 
-
-
-
-
-
+const std::map<unsigned int,unsigned int> HeatTransferModel::ModelEqData::bc_type_conversion = {
```
Comment (jbrezmorf) [inline src/transport/heat_model.cc]: Why this conversion map?

File: src/transport/concentration_model.cc [hunk 2]
```diff
@@ -42,7 +42,39 @@ using namespace std;
 using namespace Input::Type;
 
 
-
+const std::map<unsigned int,unsigned int> ConcentrationTransportModel::ModelEqData::bc_type_conversion = {
```
Comment (jbrezmorf) [inline src/transport/concentration_model.cc]: Why this conversion map?

File: src/transport/concentration_model.cc [hunk 3]
```diff
@@ -50,12 +82,38 @@ using namespace Input::Type;
 ConcentrationTransportModel::ModelEqData::ModelEqData()
 : TransportBase::TransportEqData()
 {
+    *this+=bc_type
+            .name("bc_type")
+            .description(
+            "Boundary condition type, possible values: inflow, dirichlet, neumann, robin.")
```
Comment (jbrezmorf) [inline src/transport/concentration_model.cc]: Is this correct?

File: src/transport/heat_model.cc [hunk 4]
```diff
@@ -300,6 +364,67 @@ void HeatTransferModel::compute_dirichlet_bc(const std::vector<arma::vec3> &poin
...
+	for (unsigned int i=0; i<np; i++)
+	{
+		bc_flux[i] = bc_flux_scalar[i];
+		bc_ad_value[i] = bc_ad_temperature[i];
+		bc_sigma[i] = bc_robin_sigma[i];
+		bc_ref_value[i] = bc_temperature[i];
+	}
+}
+
+void HeatTransferModel::get_diffusive_flux_bc_data(const std::vector<arma::vec3> &point_list,
```
Comment (jbrezmorf) [inline src/transport/heat_model.cc]: Do not try to make fields of heat model scalar at this time. This makes lot of boiler plate and could be probably avoided when we use multifields (we have to do that soon, should discus it).

File: src/transport/advection_diffusion_model.hh [hunk 5]
```diff
@@ -85,16 +93,80 @@ public:
...
+			arma::uvec &bc_types) = 0;
+
+	/**
 	 * Computes the Dirichlet boundary condition values.
 	 * @param point_list   Points at which to evaluate.
 	 * @param ele_acc      Element accessor.
 	 * @param bc_values    Vector of b.c. values (output).
 	 */
-	virtual void compute_dirichlet_bc(const std::vector<arma::vec3> &point_list,
+	virtual void get_dirichlet_bc_data(const std::vector<arma::vec3> &point_list,
```
Comment (jbrezmorf) [inline src/transport/advection_diffusion_model.hh]: Why we have virtual functions in the Model. The Idea to make TransportDG templated by the model was to aviod virtual calls.

**Comment by jbrezmorf** [comment]
Conclusions:
- I rather go through the documentation together.
- Need to discuss usage of multifields and how it can help in transport.
- Have to find way how to minimize boiler plate code. Choose wisely interface of Model.

File: src/transport/heat_model.cc (hunk 1 above)
Comment (jstebel) [inline src/transport/heat_model.cc]: For the moment it seems redundant, but this allows definition of different bc types (names) for Model classes. We could have e.g. "insulated_wall" in heat transfer or "precipitation" in solute transport.

File: src/transport/concentration_model.cc (hunk 2 above)
Comment (jstebel) [inline src/transport/concentration_model.cc]: See comment to heat_model.cc above.

File: src/transport/concentration_model.cc (hunk 3 above)
Comment (jstebel) [inline src/transport/concentration_model.cc]: Right, I will fix this.

File: src/transport/advection_diffusion_model.hh (hunk 5 above)
Comment (jstebel) [inline src/transport/advection_diffusion_model.hh]: Virtual methods are just for declaration of the common interface. TransportDG does not use virtual calls because it accesses directly the methods from Model.

**Comment by jstebel** [comment]
All comments to the pull request incorporated - to be merged.

File: src/transport/concentration_model.hh [hunk 6]
```diff
@@ -45,8 +45,24 @@ public:
...
-		BCField<3, FieldValue<3>::Vector> bc_conc;
+		enum Concentration_bc_types {
+			bc_inflow,
+			bc_dirichlet,
+			bc_total_flux,
+			bc_diffusive_flux
+		};
+
+		/// Convert bc_type in model to bc_type in abstract setting.
+		static const std::map<unsigned int,unsigned int> bc_type_conversion;
```
Comment (jbrezmorf) [inline src/transport/concentration_model.hh]: I think we can remove this map in order to don't be tempted to use it 
in get_bc_type. I prefer to use an lookup vector instead .

**Comment by jbrezmorf** [comment]
Remove map and merge yourself.
